### PR TITLE
Fix parameter rosette_meas_types type in docstrings

### PR DESCRIPTION
**Pull request by WayneDroid**
- [ x] This contribution adheres to [CONTRIBUTING.md](https://github.com/ni/nidaqmx-python/blob/master/CONTRIBUTING.md).

### What does this Pull Request accomplish?

Added `'python_codegen_method': 'CustomCode'` to `CreateAIRosetteStrainGageChan` in order to update parameter `rosette_meas_types` type in the function.

### Why should this Pull Request be merged?

This PR is to address issue #483.

### What testing has been done?

image

### CODE REVIEW

File: generated/nidaqmx/_task_modules/ai_channel_collection.py [hunk 1]
```diff
@@ -57,6 +56,79 @@ def _create_chan(self, physical_channel, name_to_assign_to_channel=''):
...
+                measurement.
+        Returns:
+            nidaqmx._task_modules.channels.ai_channel.AIChannel:
+
+            Indicates the newly created channel object.
+        """
+        if rosette_meas_types is None:
+            rosette_meas_types = []
+
+        rosette_meas_types = numpy.int32(rosette_meas_types)
```
Comment (zhindes) [inline generated/nidaqmx/_task_modules/ai_channel_collection.py:120]: It's not just the docstring that is wrong. This is wrong, too. This should be converting the `.value` of the enumerations. The way to validate your change is to update [this test](https://github.com/ni/nidaqmx-python/blob/a7a3643911fd549cac198c0a2cc8dadd4a8ad8f4/tests/component/_task_modules/channels/test_ai_channel.py#L848) to not create an array of `.value`, just the enumerations themselves.

**Review by zhindes — CHANGES_REQUESTED** [review]
This is either a gap in our support for arrays of enumerations or some sort of issue with this specific function definition in the metadata. I want to say we have array of enums somewhere else in the API, so I think maybe the metadata is bad?

**Comment by zhindes** [comment]
> This is either a gap in our support for arrays of enumerations or some sort of issue with this specific function definition in the metadata. I want to say we have array of enums somewhere else in the API, so I think maybe the metadata is bad?

My point here is I don't think just no longer code-generating rosette is correct. I think there is a metadata problem.

File: generated/nidaqmx/_task_modules/ai_channel_collection.py (hunk 1 above)
Comment (bkeryan) [reply generated/nidaqmx/_task_modules/ai_channel_collection.py:120]: Note that the way this uses `numpy.int32(some_list)` to create a NumPy array is not well documented and `numpy.typing` doesn't allow this behavior. If we touch any code like this, let's update it to use `numpy.array(some_list, dtype=numpy.int32)`.

**Comment by bkeryan** [comment]
> > This is either a gap in our support for arrays of enumerations or some sort of issue with this specific function definition in the metadata. I want to say we have array of enums somewhere else in the API, so I think maybe the metadata is bad?
> 
> My point here is I don't think just no longer code-generating rosette is correct. I think there is a metadata problem.

I think `rosette_meas_types = numpy.int32(rosette_meas_types)` is generated based on `get_instantiation_lines`:
https://github.com/ni/nidaqmx-python/blob/a7a3643911fd549cac198c0a2cc8dadd4a8ad8f4/src/codegen/utilities/function_helpers.py#L136

It has a special case for enum arrays, so yeah, I think the metadata for this parameter is wrong.

File: generated/nidaqmx/_task_modules/ai_channel_collection.py (hunk 1 above)
Comment (WayneDroid) [reply generated/nidaqmx/_task_modules/ai_channel_collection.py]: @zhindes  @bkeryan I updated functions.py to have `rosette_meas_types` parameter reflected with correct type, I'm not sure is this correct approach. If it's okay, I will continue to work on scrapigen to update the metadata accordingly.

File: src/codegen/templates/_task_modules/ai_channel_collection.py.mako [hunk 2]
```diff
@@ -37,8 +37,8 @@ class AIChannelCollection(ChannelCollection):
             name_to_assign_to_channel (Optional[str]): Specifies a name to
                 assign to the virtual channel this method creates.
         Returns:
-            nidaqmx._task_modules.channels.ai_channel.AIChannel: 
-            
+            nidaqmx._task_modules.channels.ai_channel.AIChannel:
+
             Specifies the newly created AIChannel object.
```
Comment (bkeryan) [inline src/codegen/templates/_task_modules/ai_channel_collection.py.mako:42]: Adding a blank line between the return type and return description looks wrong. I guess it's moving the 1st line to the 2nd paragraph. I know some of the other files in this repo do this too but that doesn't make it better than what was there before.

https://www.sphinx-doc.org/en/master/usage/extensions/example_google.html 

BTW this method is private so it doesn't really matter how we document it.

**Review by bkeryan — APPROVED** [review]
Approved with suggestions (revert doc change or delete unnecessary docstring)

Sorry, the review was in pending for a couple days because I forgot to submit it.

File: src/codegen/templates/_task_modules/ai_channel_collection.py.mako (hunk 2 above)
Comment (WayneDroid) [reply src/codegen/templates/_task_modules/ai_channel_collection.py.mako:42]: The blank line between return type and description has been there in current file ([ai_channel_collection.py.mako](https://github.com/ni/nidaqmx-python/blob/205dc3bcd6f5c32c7d4f1ed46e66af30af65f927/src/codegen/templates/_task_modules/ai_channel_collection.py.mako#L41)). Based on the diff-view on generated ai_channel_collection.py, I believe it's my code editor removed whitespaces at line 41 on mako file when I saved the file after reverting the changes from first commit.
image

I find that other public methods having a blank line between return type and description, maybe I should keep the blank line to align with existing docstring format.

Comment (bkeryan) [reply src/codegen/templates/_task_modules/ai_channel_collection.py.mako:42]: Oh, I see. Removing trailing whitespace without changing the formatting is fine.

**Review by maxxboehme — APPROVED** [review]

File: generated/nidaqmx/_task_modules/ai_channel_collection.py [hunk 3]
```diff
@@ -1503,13 +1503,15 @@ def add_ai_rosette_strain_gage_chan(
...
+
             Indicates the newly created channel object.
         """
         if rosette_meas_types is None:
             rosette_meas_types = []
 
-        rosette_meas_types = numpy.int32([p.value for p in rosette_meas_types])
+        temp_vals = [p.value for p in rosette_meas_types]
+
+        rosette_meas_types = numpy.int32(temp_vals)
```
Comment (WayneDroid) [inline generated/nidaqmx/_task_modules/ai_channel_collection.py:1514]: @bkeryan  @zhindes Help needed for the mypy check error. 

image

I tried type hint `List[StrainGageRosetteMeasurementType]` to `rosette_meas_types`, but I still get the similar error. Any suggestions? 
image
image

image

Comment (zhindes) [reply generated/nidaqmx/_task_modules/ai_channel_collection.py:1514]: I think the solution is to type `temp_vals` like this:


```python
temp_vals: List[int] = [p.value for p in rosette_meas_types]
```

**Review by zhindes — APPROVED** [review]

File: generated/nidaqmx/_task_modules/ai_channel_collection.py (hunk 3 above)
Comment (bkeryan) [reply generated/nidaqmx/_task_modules/ai_channel_collection.py:1514]: Rerunning codegen on your branch does this:
```
-        temp_vals = [p.value for p in rosette_meas_types]
-
-        rosette_meas_types = numpy.int32(temp_vals)
+        rosette_meas_types = numpy.int32([p.value for p in rosette_meas_types])
```

Which tells me two things:
- I missed one of the codegen templates
- build.yml should check for dirty codegen *before* it runs linters :)

Comment (bkeryan) [reply generated/nidaqmx/_task_modules/ai_channel_collection.py:1514]: You need to update this: https://github.com/ni/nidaqmx-python/blob/master/src/codegen/utilities/function_helpers.py#L133

If `param.ctypes_data_type.startswith("numpy.")` then it should generate `numpy.array({1}, dtype={0})`, and likewise for the array version.

Comment (WayneDroid) [reply generated/nidaqmx/_task_modules/ai_channel_collection.py]: @bkeryan @zhindes Thanks for the suggestions.

For some reasons, I can't use `numpy.array({1}, dtype={0})` for `rosette_meas_types` (enum list). As a generated code `rosette_meas_types = numpy.array(rosette_meas_types, dtype=numpy.int32)` causing pytest `Type Error` failure. I have to extract value from enum list with `numpy.array([p.value for p in {0}], dtype={1})`.

image

Comment (bkeryan) [reply generated/nidaqmx/_task_modules/ai_channel_collection.py]: That's right. I should have said "and likewise for the enum-array version"

**Review by bkeryan — APPROVED** [review]

**Pull request merged**